### DISCUSSION TITLE
Why is `FullTextType` compared as string?

### DISCUSSION THREAD

**Discussion by lsndr** [mikro-orm/mikro-orm#5908] [Q&A] [answered]
For some reason `FullTextType` is configured to be compared as string:

https://github.com/mikro-orm/mikro-orm/blob/05285ed378a5644ae0612cdf8d56a04988ec7849/packages/postgresql/src/types/FullTextType.ts#L19-L21

But `convertToDatabaseValue` might also return `RawQueryFragment`, which makes string comparison useless, since 2 different objects can't be equal:

https://github.com/mikro-orm/mikro-orm/blob/05285ed378a5644ae0612cdf8d56a04988ec7849/packages/postgresql/src/types/FullTextType.ts#L30

Here is how strings compared:

https://github.com/mikro-orm/mikro-orm/blob/05285ed378a5644ae0612cdf8d56a04988ec7849/packages/core/src/utils/EntityComparator.ts#L656-L658

Such comparison behavior makes entity to be saved every time it's loaded from database, even if it's not modified.

Is it intended behavior or a bug?

**Comment by B4nan** [comment]
> Is it intended behavior or a bug?

The answer to this is yet another question. Do you have a failing repro or is this just a thought? :]

This is a community contribution, so I am not entirely sure, but I believe you are not supposed to be changing the value anyhow, such properties are meant for reading/searching, you should have a regular string property and update the fulltext column value via `onUpdate` property hook as shown in the docs.

https://mikro-orm.io/docs/query-conditions#postgresql

**Reply by lsndr** [reply]
Here is the reproduction: I don't modify entity in the code anyhow, so it should not get updated 

https://github.com/lsndr/mikroorm-bug-7-aug

**Reply by B4nan** [reply]
Hmm, that's indeed weird, let me check the existing tests, maybe they behave the same. Note that changing the `compareAsType` won't help here, it's not about that most likely.

**Reply by lsndr** [reply]
As workaround I will probably use this custom type:

```typescript
class FullTextType extends BaseFullTextType {
  private convertToScalar(value: null | string | RawQueryFragment) {
    return value instanceof RawQueryFragment ? `${value.sql};${value.params.join(',')}` : value;
  }

  override compareValues(a: null | string | RawQueryFragment, b: null | string | RawQueryFragment): boolean {
    return this.convertToScalar(a) === this.convertToScalar(b);
  }
}
```

I also noticed, that adding `lazy: true` to the `nameVector` in the provided reproduction helps to avoid wrong comparison. But for some reason lazy loading doesn't work in my real application (trying to figure out why)

**Reply by B4nan** [reply]
Yeah, old tests are also suffering from this extra update.

**Reply by lsndr** [reply]
> Yeah, old tests are also suffering from this extra update.

Should I create an issue?

Also I figured out why `lazy: true` doesn't work in my case. If I load entity using `createQueryBuilder`, lazy loading doesn't work:

```typescript
const entity = await em.createQueryBuilder(DogEntity).getSingleResult();
```

Is it expected?

**Reply by B4nan** [reply]
you need to be more specific, wdym by doesnt work?

**Reply by lsndr** [reply]
According to the [documentation](https://mikro-orm.io/docs/defining-entities#lazy-scalar-properties), fields, which are marked with `lazy: true` are omitted from select clause except when they are explicitly specified via `populate` option.

The above is true, if I use EM's `find` and `findOne` methods. But if I use `createQueryBuilder` as I described in the previous comment, all fields are included in the select clause:

```sql
select "d0".* from "dog_entity" as "d0" limit 1
``` 

If you need, I can create a reproduction repo for this

**Reply by B4nan** [reply]
With QB you need to be explicit, `lazy` flag is for entity manager, I don't consider this as a bug.

**Reply by B4nan** [reply, marked as answer]
This will help with the extra updates d18e5ede53886abdf32ac036e41a3bc4e8f293cf
